### ISSUE TITLE
trainer/grpo: GRPO objective variants silently fall back to stock TRL loss due to missing per_token_logps in inputs

### ISSUE THREAD

**Issue by here-2007** [MakazhanAlpamys/Soup#584]
### Description

In `src/soup_cli/trainer/grpo.py`, `_GRPOTrainerVariant.compute_loss()` attempts to read `per_token_logps` directly from `inputs`:

```python
logp_new = _read_attr(inputs, "per_token_logps")
logp_old = _read_attr(inputs, "old_per_token_logps")
if logp_old is None:
    logp_old = _read_attr(inputs, "ref_per_token_logps")
advantages = _read_attr(inputs, "advantages")
completion_mask = _read_attr(inputs, "completion_mask")

if logp_new is None or logp_old is None or advantages is None:
    # Fall back to the original loss — defence-in-depth so a
    # TRL internal rename does not crash the training loop.
    self._warn_fallback("missing per-token log-prob inputs")
    return super().compute_loss(model, inputs, return_outputs=return_outputs, **kwargs)
```

However, in `trl.GRPOTrainer`, `inputs` delivered to `compute_loss()` (from `_generate_and_score_completions`) only contains dataset columns, token IDs, masks, and advantages (`prompt_ids`, `prompt_mask`, `completion_ids`, `completion_mask`, `advantages`, `num_items_in_batch`).

`per_token_logps` represents the trainable policy log-probabilities requiring active gradients, and is **never** pre-populated in `inputs` before `compute_loss()`. In TRL, `per_token_logps` is computed inside `_compute_loss()` via `self._get_per_token_logps_and_entropies(...)`.

Because `logp_new` is always `None` when `compute_loss()` starts:
1. `if logp_new is None or logp_old is None or advantages is None` evaluates to `True` on every training step.
2. `self._warn_fallback("missing per-token log-prob inputs")` is triggered.
3. `super().compute_loss(...)` is returned.

**Impact:** Every configured GRPO variant (`gspo`, `dapo`, `dr_grpo`, `bnpo`, `two_sided`, `rft`) silently falls back to standard stock TRL GRPO loss. The selected variant objective is never applied during actual model training.

---

### Steps to reproduce

1. Configure a GRPO training run with any variant (e.g. `gspo`):
```yaml
task: grpo
training:
  grpo_variant: gspo
  learning_rate: 1e-5
  num_generations: 4
  max_completion_length: 128
```
2. Start training with `soup train --config config.yaml`.
3. Observe the trainer log output during the first training step.

**Expected behavior:**
The trainer computes the GSPO objective loss via `apply_variant_loss("gspo", ...)` and optimizes the model using group-stabilized policy optimization.

**Actual behavior:**
The one-shot fallback warning is emitted:
```text
WARNING: GRPO variant 'gspo' compute_loss fell back to the stock TRL loss (missing per-token log-prob inputs); the selected objective is NOT being applied. This usually means a TRL version renamed the per-token log-prob inputs.
```
and standard stock TRL GRPO loss is used for the entire training run.

---

### Root Cause & Suggested Solution

In `src/soup_cli/trainer/grpo.py`:
1. When `per_token_logps` is not pre-populated in `inputs`, reconstruct `input_ids` and `attention_mask` from `prompt_ids` and `completion_ids`, and obtain `per_token_logps` in a single forward pass via `self._get_per_token_logps_and_entropies(model, ...)`.
2. Fall back to `per_token_logps.detach()` for `logp_old` when `old_per_token_logps` is not present (matching TRL's single-iteration behavior).
3. Route loss computation through `_compute_loss(self, model, inputs)` and `compute_loss(self, model, inputs)` so variant loss is returned with proper gradient accumulation normalization without executing redundant forward passes or increasing peak VRAM.

---

### Environment

- **OS:** macOS / Linux / Windows
- **Soup version:** `0.73.3` (main branch)
- **TRL version:** `0.29.1` (satisfies `trl>=0.29.0,<1.0.0`)
- **Python version:** `3.10` / `3.11` / `3.12`

**Comment by here-2007** [comment]
Taking this issue myself.

**Comment by MakazhanAlpamys** [comment]
Closed by merged PR #585 (`47b9121`). Credit to @here-2007 for routing GRPO variants through TRL per-token log-probability computation, preserving detached old probabilities, and adding mutation-backed runtime-contract coverage.
